### PR TITLE
feat: add dot command to GitUICommands

### PR DESCRIPTION
**Pull request by MrJithil**
Fixes #

## Proposed changes

- dot command is widespread in most applications to open the current working directory. I have added a `private` function to handle the `.` command in this pull request. For Eg: If a user wants to open their current repository in GitExtensions, he should trigger the command `GitExtensions .`

## Screenshots 
DotCommand

### Before
A user needs to enter different commands to open his current repository.
`GitExtensions browse .` is an example.

### After
By these changes, a user can open the repository in GitExtensions by just triggering the command `GitExtensions .`

### TODO
Need documentation for this command once this PR got sufficient approvals. 

## Test methodology 

- I have tested the changes by modifying the project launch setting property.
- Also, to ensure functionality, I have run the executables from a different git repository.
- Both negative and positive cases were tested.
- Since the added function is a private one, we don't need any unit tests.
- I captured the evidence from my local machine in this pull request.

## Test environment(s) 

- GIT 
- Windows 

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/GitUICommands.cs
Comment (gerhardol) [inline GitUI/GitUICommands.cs:1474]: see comment on line 1383: Update FormCommandlineHelp too

Comment (gerhardol) [inline GitUI/GitUICommands.cs:1474]: Should this be handled in default:?
A specific case for handling paths.

**Review by gerhardol — COMMENTED** [review]
seem ok
The normal way is to use "browse <path>" or "gitex.cmd <path>"

File: GitUI/GitUICommands.cs
Comment (MrJithil) [reply GitUI/GitUICommands.cs]: Yes, I have added that. Thanks for the knowledge share.

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: Sure. Thanks for the great suggestion.
I have added that.

**Comment by MrJithil** [comment]
@gerhardol I have added the suggested changes. Please review.

File: GitUI/GitUICommands.cs
Comment (mstv) [inline GitUI/GitUICommands.cs:1926]: Please remove. I do not see any benefit of adding an alias. Furthermore "." is rather a path than a command.

**Review by mstv — COMMENTED** [review]
> The normal way is to use "browse " or "gitex.cmd "

So if we start to support ".", why not any path?
I don't know the rationale why the usage of `browse` was forced when the CLI was designed.

File: GitUI/GitUICommands.cs
Comment (mstv) [reply GitUI/GitUICommands.cs]: The `default:` case is dealing with clone URLs.
Without a comment, the added code will be hard to understand next time we will read it - basically, i.e. not only if not handled using `case ".":`.

File: GitUI/GitUICommands.cs
Comment (gerhardol) [reply GitUI/GitUICommands.cs:1926]: Agree. No alias, add a comment that this is interpreted as a path
Also add a test, similar to RunCommandBasedOnArgument_browse()

File: GitUI/GitUICommands.cs
Comment (MrJithil) [reply GitUI/GitUICommands.cs]: I have added the comment as discussed. The changes will be reflected in the next push.

File: GitUI/GitUICommands.cs
Comment (MrJithil) [reply GitUI/GitUICommands.cs:1926]: Sure. I have added the test case `RunCommandBasedOnArgument_dot()`.
Thank you much for the sample reference.
The changes will be in the next pull.
CC: @mstv @gerhardol

**Review by gerhardol — APPROVED** [review]

File: IntegrationTests/UI.IntegrationTests/GitUICommands/RunCommandTests.cs
Comment (mstv) [inline IntegrationTests/UI.IntegrationTests/GitUICommands/RunCommandTests.cs:372]: What is the purpose of those unused instances?

File: GitUI/GitUICommands.cs
Comment (mstv) [inline GitUI/GitUICommands.cs:1494]: If we start to support ".", why not any path? Thoughts, @gitextensions/git-extensions-team?

Then `Program.GetWorkingDir` should try `args[1]`, too. Currently, it is hard-coded to `args[2]`.

**Review by mstv — COMMENTED** [review]
@MrJithil, file another PR with your sign off in `contributors.txt`, please.

File: IntegrationTests/UI.IntegrationTests/GitUICommands/RunCommandTests.cs
Comment (MrJithil) [reply IntegrationTests/UI.IntegrationTests/GitUICommands/RunCommandTests.cs]: Thought to extend the unit tests. But, since it's handled with a function, removed it.

Comment (MrJithil) [reply IntegrationTests/UI.IntegrationTests/GitUICommands/RunCommandTests.cs]: Removed in the latest push.

**Comment by MrJithil** [comment]
> @MrJithil, file another PR with your sign off in `contributors.txt`, please.

I have created #9148 as my sign-off. Thanks @mstv

File: GitUI/GitUICommands.cs
Comment (gerhardol) [reply GitUI/GitUICommands.cs:1494]: Why not?

Comment (RussKie) [reply GitUI/GitUICommands.cs:1494]: I'm with @mstv - "." denotes a path, so its handling should be generic to any path.

Comment (MrJithil) [reply GitUI/GitUICommands.cs:1494]: So, do we need to create a separate PR for that? Or do I need to make changes to this one?
CC: @mstv

Comment (MrJithil) [reply GitUI/GitUICommands.cs:1494]: Anyway, I started working on it.

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: Implemented it.

**Comment by MrJithil** [comment]
@mstv @gerhardol @RussKie I have implemented the suggestion to take care `path` along with the `.` param.

Please see the evidence here. Kindly review, approve and merge if you are okay to proceed. Please feel free to revert back in case of any additional suggestions.

New 

Evidence:


https://user-images.githubusercontent.com/26359740/117197332-52135400-ae05-11eb-9c31-904b6c76ead9.mp4

File: GitUI/CommandsDialogs/FormCommandlineHelp.resx
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormCommandlineHelp.resx:121]: remove .

File: GitUI/GitUICommands.cs
Comment (gerhardol) [inline GitUI/GitUICommands.cs:1493]: Suggest reducing the code (untested)
```suggestion
                        List<string> args2 = new() { "browse" };
                        args2.AddRange(args);
                        return RunBrowseCommand(args2);
```

**Review by gerhardol — COMMENTED** [review]
+1

File: GitUI/CommandsDialogs/FormCommandlineHelp.resx
Comment (mstv) [reply GitUI/CommandsDialogs/FormCommandlineHelp.resx:121]: Or rather turn into
```suggestion
    <value>[path]
```
?

File: GitUI/GitUICommands.cs
Comment (mstv) [reply GitUI/GitUICommands.cs:1493]: I am afraid `RunBrowseCommand` will not re-evaluate the path because it has already been checked and fed to the `Module`.

`RunOpenRepoFromPath` is less flexible than the existing logic in `Program.GetWorkingDir`, which should be re-used definitely.
I suggest to just call `LaunchBrowse(args[1])` - also if `File.Exists(args[1])`.
This has the drawback that a new instance is started. But  this should not be too expensive nowadays.

Comment (MrJithil) [reply GitUI/GitUICommands.cs:1493]: We may not do this approach for reducing the code. Because `browse` or `openrepo` commands are trying to open a repository that is set during the instantiation (`GitModule`). 
So, instead of proceeding with a workaround, my opinion is to treat it as a separate function as now.

Please suggest if you have disagreements.

File: GitUI/CommandsDialogs/FormCommandlineHelp.resx
Comment (MrJithil) [reply GitUI/CommandsDialogs/FormCommandlineHelp.resx:121]: Removed.

File: GitUI/GitUICommands.cs
Comment (mstv) [reply GitUI/GitUICommands.cs:1493]: > So, instead of proceeding with a workaround, my opinion is to treat it as a separate function as now.

I disagree to duplicating or - even worse - creating slightly different logic.

You can ignore the returned `IProcess` instance (as all other callers do, too). The new GE instance shall run detached.
Just `return true` in order to exit the first GE instance.

Comment (gerhardol) [reply GitUI/GitUICommands.cs:1493]: mstv suggestion avoids the additional method
(I will not force checking if args[1] is a file to use the directory).
```suggestion
LaunchBrowse(args[1]);
```

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: Sure. I have added those changes. I kept the Directory Exisist checks there, to not break the code and avoid the malformed commands. 

@mstv @gerhardol Please review.

**Review by RussKie — APPROVED** [review]
Works for me 👍

File: GitUI/GitUICommands.cs
Comment (mstv) [inline GitUI/GitUICommands.cs:1491]: Browse supports files, too. Do you have an example where we would not see a relevant error message when using:
```suggestion
                    if (args.Length == 2 && (Directory.Exists(args[1]) || File.Exists(args[1])))
```
?

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: `LaunchBrowse` only supports File. So, get the directory if it's a path, then open it up.

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: Added

File: GitUI/GitUICommands.cs
Comment (gerhardol) [inline GitUI/GitUICommands.cs:1494]: Exception if no argument...
untested

```suggestion
                    string dir = args.Count != 2 
                        ? ""
                        :  File.Exists(args[1]) ? Path.GetDirectoryName(args[1]) : args[1];
                    if (Directory.Exists(dir))
                    {
                        LaunchBrowse(dir);
```

File: GitUI/GitUICommands.cs
Comment (mstv) [reply GitUI/GitUICommands.cs]: > `LaunchBrowse` only supports File. So, get the directory if it's a path, then open it up.

Why do you think this? It supports the repo dir as well as any dir or file below. Just play around with `GitExtensions.exe browse <path>`. It opens the repo. The tail is ignored.

Comment (mstv) [reply GitUI/GitUICommands.cs]: Ah, I forgot about the (arguable) design of `.Exists`. So exceptions from the `.Exists` functions must be silently caught.

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: The same changes already added in the MR. please test it if necessary

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: And, I request the approval as the earliest since its a been a while.

Comment (mstv) [reply GitUI/GitUICommands.cs]: > Exception if no argument...

as @gerhardol wrote.
Avoid too much logic, just ignore the exceptions but catch them.

File: GitUI/GitUICommands.cs
Comment (MrJithil) [reply GitUI/GitUICommands.cs:1494]: I don't see any exception if I am passing no args. Could you please give evidence?

Since we are `string.IsNullOrWhiteSpace(args[1])` invalid cases will be ignored.

File: GitUI/GitUICommands.cs
Comment (MrJithil) [reply GitUI/GitUICommands.cs]: We will not get any exceptions. Also, I just add only one line code to avoid a try and catch. It will not even affect the complete flow.
Do you think, the try-catch is necessary here? If yes, do we need an extra `function` to handle it? Because the if condition will be exhausted with too many lines.

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: @mstv

File: GitUI/GitUICommands.cs
Comment (RussKie) [inline GitUI/GitUICommands.cs:1491]: [nit] put `args[1]` into a variable, and use it across.

Comment (MrJithil) [reply GitUI/GitUICommands.cs:1491]: We discussed reducing the code before. Now, are we okay to add more lines?
Please confirm.

Comment (MrJithil) [reply GitUI/GitUICommands.cs]: I appreciate it. Instead of creating a variable for `args[1]`, I reused the variable `command` here.

**Comment by RussKie** [comment]
Please [squash](https://github.com/gitextensions/gitextensions/wiki/How-To%3A-Squash-and-Rebase-your-changes) and it is ready to go.

**Comment by MrJithil** [comment]
> Please [squash](https://github.com/gitextensions/gitextensions/wiki/How-To%3A-Squash-and-Rebase-your-changes) and it is ready to go.

Rebased the branch with `master`. @RussKie Please proceed.

**Comment by RussKie** [comment]
Thank you

**Pull request merged**
